src/compiler/checker.ts

```diff
@@ -9912,14 +9921,10 @@ namespace ts {
                 emitAwaiter = true;
             }
 
-            const returnType = node.type && getTypeFromTypeNode(node.type);
-            let promisedType: Type;
-            if (returnType && isAsync) {
-                promisedType = checkAsyncFunctionReturnType(node);
-            }
-
-            if (returnType && !node.asteriskToken) {
-                checkAllCodePathsInNonVoidFunctionReturnOrThrow(node, isAsync ? promisedType : returnType);
+            const returnOrPromisedType = node.type && (isAsync ? checkAsyncFunctionReturnType(node) : getTypeFromTypeNode(node.type));

```

Comment:
Looks like a couple of places where we have to choose between `checkAsyncFunctionReturnType` and `getTypeFromTypeNode`. I think it would be a good idea to have a `getFunctionReturnType` that just wraps the decision. (And it's quite possible that additional calls to `getTypeFromTypeNode` need to switch.)